File: baekjoon/14499.py
# input
# 4 2 0 0 8
# 0 2
# 3 4
# 5 6
# 7 8
# 4 4 4 1 3 3 3 2
# output
# 0
# 0
# 3
# 0
# 0
# 8
# 6
# 3

N, M, x, y, K = map(int, input().split(" "))

class Dice:
    def __init__(self):
        self.top = 0
        self.bottom = 0
        self.east = 0
        self.west = 0
        self.south = 0
        self.north = 0
    
    def move(self, num):
        if num == 1:
            self.go_east()
        elif num == 2:
            self.go_west()
        elif num == 3:
            self.go_north()
        else:
            self.go_south()
    
    def go_east(self):
        self.top, self.bottom, self.east, self.west = self.west, self.east, self.top, self.bottom
    
    def go_west(self):
        self.top, self.bottom, self.east, self.west = self.east, self.west, self.bottom, self.top
    
    def go_north(self):
        self.top, self.bottom, self.south, self.north = self.south, self.north, self.bottom, self.top

    def go_south(self):
        self.top, self.bottom, self.south, self.north = self.north, self.south, self.top, self.bottom

dice = Dice()

board = [list(map(int, input().split(" "))) for _ in range(N)]

com = list(map(int, input().split(" ")))

dx, dy = [0, 0, -1, 1], [1, -1, 0, 0]

curr_loc = [x, y]

for c in com:
    xx, yy = curr_loc[0]+dx[c-1], curr_loc[1]+dy[c-1]
    if not (0 <= xx < N) or not (0 <= yy < M):
        continue
    curr_loc = [xx, yy]
    dice.move(c)
    if board[xx][yy] == 0:
        board[xx][yy] = dice.bottom
    elif board[xx][yy] != 0:
        dice.bottom = board[xx][yy]
        board[xx][yy] = 0
    print(dice.top)

File: baekjoon/gold/14499/14499.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

var (
	reader *bufio.Reader = bufio.NewReader(os.Stdin)
	writer *bufio.Writer = bufio.NewWriter(os.Stdout)
	dx                   = []int{0, 0, 0, -1, 1}
	dy                   = []int{0, 1, -1, 0, 0}
)

type Dice struct {
	top, bottom, north, south, east, west int
}

func (d *Dice) move(dir int) {
	switch dir {
	case 1: // east
		d.top, d.bottom, d.east, d.west = d.west, d.east, d.top, d.bottom
	case 2: // west
		d.top, d.bottom, d.east, d.west = d.east, d.west, d.bottom, d.top
	case 3: // north
		d.top, d.bottom, d.north, d.south = d.south, d.north, d.top, d.bottom
	case 4: // south
		d.top, d.bottom, d.north, d.south = d.north, d.south, d.bottom, d.top
	}
}

func NewDice() *Dice {
	return &Dice{0, 0, 0, 0, 0, 0}
}

func main() {
	defer writer.Flush()

	var N, M, x, y, K int
	var board [][]int

	fmt.Fscanf(reader, "%d %d %d %d %d\n", &N, &M, &x, &y, &K)
	board = make([][]int, N)
	for i := range board {
		nums, _ := reader.ReadString('\n')
		nums = strings.TrimSpace(nums)
		board[i] = splitInt(nums)
	}
	nums, _ := reader.ReadString('\n')
	nums = strings.TrimSpace(nums)
	commands := splitInt(nums)
	dice := NewDice()

	for _, cmd := range commands {
		nx, ny := x+dx[cmd], y+dy[cmd]
		if nx < 0 || nx >= N || ny < 0 || ny >= M {
			continue
		}
		x, y = nx, ny
		dice.move(cmd)
		if board[x][y] == 0 {
			board[x][y] = dice.bottom
		} else {
			dice.bottom = board[x][y]
			board[x][y] = 0
		}
		fmt.Fprintf(writer, "%d\n", dice.top)
	}
}

func splitInt(s string) []int {
	var res []int
	for _, v := range strings.Fields(s) {
		res = append(res, parseInt(v))
	}
	return res
}

func parseInt(s string) int {
	var n int
	fmt.Sscanf(s, "%d", &n)
	return n
}
